src/main/java/com/puppycrawl/tools/checkstyle/DefaultLogger.java

```diff
@@ -211,4 +214,93 @@ private void closeStreams() {
         }
     }
 
+    /**
+     * Represents a message that can be localised. The translations come from
+     * message.properties files. The underlying implementation uses
+     * java.text.MessageFormat.
+     */
+    private static final class LocalizedMessage {
+
+        /**
+         * A cache that maps bundle names to ResourceBundles.
+         * Avoids repetitive calls to ResourceBundle.getBundle().
+         */
+        private static final Map<String, ResourceBundle> BUNDLE_CACHE =
+                Collections.synchronizedMap(new HashMap<>());
+
+        /**
+         * The locale to localise messages to.
+         **/
+        private static Locale locale = Locale.getDefault();
+
+        /**
+         * Key for the message format.
+         **/
+        private final String key;
+
+        /**
+         * Arguments for MessageFormat.
+         */
+        private final Object[] args;
+
+        /**
+         * Creates a new {@code LocalizedMessage} instance.
+         *
+         * @param key the key to locate the translation.
+         */
+        /* package */ LocalizedMessage(String key) {
```

Comment:
`package` comment has been added as we cannot make it private due to pmd rule that private constructor cannot be used outside its class, and it cannot be public as checkstyle gives redundant scope violation.

Reply:
it is not good that static analysis tools force us to make bad design of class. PMD is right. Checkstyle is not right.
Please register issue on this Check (if not yet already created).
Please make ctor public and lets suppress violation by Xpath
https://checkstyle.org/config_filters.html#SuppressionXpathSingleFilter , nested classes should allow public scope.

Reply:
problem has bigger scope , we will do it seprately - https://github.com/checkstyle/checkstyle/issues/10161

for now package is ok.